File: docs/build_docs.py
import os
from pathlib import Path

FILE_TAG = '[example]'
CODE_BLOCK_HEADER = '```typescript { name=<name>, category=example }'
CODE_BLOCK = '```'
CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))
BUILD_DIR = os.path.join(CURRENT_DIR, "build")
TEMPLATE_DIR = os.path.join(CURRENT_DIR, "templates")

def get_snippet(file_line: str) -> str:
    """ Gets snippet text from referenced file """
    print(f"\t\tGetting snippet from: {file_line}")
    snippet_file_rel_path = file_line.replace(FILE_TAG, '').replace('(','').replace(')','').strip()
    snippet_file_abs_path = os.path.join(CURRENT_DIR, Path(snippet_file_rel_path))
    snippet_content = Path(snippet_file_abs_path).read_text()
    return snippet_content

def build_md_file(md_file_name: str):
    """ Builds individual markdown file"""
    print(f"\tBuilding file {md_file_name}")
    try:
        input_file_name = os.path.basename(md_file_name).replace('.md', '')
        snippet_counter = 0
        input_file_path = Path(md_file_name)
        output_file_path = Path(os.path.join(BUILD_DIR, os.path.basename(input_file_path)))
        input_content = input_file_path.read_text().splitlines()
        for line_index in range(0, len(input_content)):
            if FILE_TAG in input_content[line_index]:
                snippet_content = get_snippet(input_content[line_index])
                with output_file_path.open("a") as f:
                    f.write(f"{CODE_BLOCK_HEADER.replace('<name>', f'{input_file_name}{snippet_counter}')}\n")
                    f.write(f"{snippet_content}\n")
                    f.write(f"{CODE_BLOCK}\n")
                snippet_counter += 1
            else:
                with output_file_path.open("a") as f:
                    f.write(f"{input_content[line_index]}\n") 
    except Exception as e:
        print(f"\tUnable to build file {md_file_name}")
        raise e

def get_template_files() -> []:
    """ Gets list of template markdown files to process"""
    files = [f.lower() for f in os.listdir(TEMPLATE_DIR) if os.path.isfile(f) and f.endswith('.md')]
    print(f"\t{len(files)} template markdown files found")
    return files

def main():
    """ Entry point """
    print('Replacing code snippets in template files:')
    md_files = get_template_files()
    for mdfile in md_files:
        build_md_file(mdfile)

if __name__ == '__main__':
    main()

File: docs/builddocs.ts
import * as fs from 'fs';
import * as path from 'path';

const FILE_TAG = '[example]';
const CODE_BLOCK_HEADER = '```typescript { name=<name>, category=example }';
const CODE_BLOCK = '```';
const CURRENT_DIR = path.resolve();
const TEMPLATE_DIR = path.join(CURRENT_DIR, 'templates');

function getSnippetFilePath(tagLine: string): string {
    const snippetFileRelPath = tagLine
        .replace(FILE_TAG, '')
        .replace('(', '')
        .replace(')', '')
        .trim();
    const snippetFileAbsPath = path.join(CURRENT_DIR, snippetFileRelPath);
    return snippetFileAbsPath;
}

function getSnippetName(tagLine: string): string {
    const snippetFileAbsPath = getSnippetFilePath(tagLine);
    return path.basename(snippetFileAbsPath).toLowerCase().replace('.ts', '');
}

function getSnippetContent(tagLine: string): string {
    console.log(`\t\tGetting snippet from: ${tagLine}`);
    const snippetFileAbsPath = getSnippetFilePath(tagLine);
    const snippetContent = fs.readFileSync(snippetFileAbsPath, 'utf8');
    return snippetContent;
}

function buildTemplate(templateFileName: string): void {
    console.log(`\tBuilding file ${templateFileName}`);
    try {
        const templateFilePath = path.join(TEMPLATE_DIR, templateFileName);
        const outputFilePath = path.join(CURRENT_DIR, templateFileName);
        const templateContent = fs
            .readFileSync(templateFilePath, 'utf8')
            .split('\n');
        for (
            let lineIndex = 0;
            lineIndex < templateContent.length;
            lineIndex++
        ) {
            if (templateContent[lineIndex].includes(FILE_TAG)) {
                const tagLine = templateContent[lineIndex];
                const snippetContent = getSnippetContent(tagLine);
                const snippetName = getSnippetName(tagLine);
                fs.appendFileSync(
                    outputFilePath,
                    `${CODE_BLOCK_HEADER.replace('<name>', snippetName)}\n`
                );
                fs.appendFileSync(outputFilePath, `${snippetContent}\n`);
                fs.appendFileSync(outputFilePath, `${CODE_BLOCK}\n`);
            } else {
                fs.appendFileSync(
                    outputFilePath,
                    `${templateContent[lineIndex]}\n`
                );
            }
        }
    } catch (e) {
        console.log(`\tUnable to build template ${templateFileName}`);
        throw e;
    }
}

function getTemplateFiles(): string[] {
    const files = fs
        .readdirSync(TEMPLATE_DIR)
        .filter((f) => f.toLowerCase().endsWith('.md'));
    console.log(`\t${files.length} template markdown files found`);
    return files;
}

function main(): void {
    console.log('Replacing code snippets in template files:');
    const mdTemplates = getTemplateFiles();
    for (const mdTemplate of mdTemplates) {
        buildTemplate(mdTemplate);
    }
}

main();
